pkg/tagset/caching_factory.go

```diff
@@ -0,0 +1,179 @@
+		panic("cacheWidth must be at least 1")
+	}
+
+	// approximate the tagsCache parameters, rounding up
+	insertsPerRotation := cacheSize/cacheWidth + 1
+	cacheCount := cacheWidth
+
+	var caches [numCacheIDs]tagsCache
+	for i := range caches {
+		caches[i] = newTagsCache(insertsPerRotation, cacheCount)
+	}
+	return &cachingFactory{
+		caches: caches,
+	}
+}
+
+// NewTags implements Factory.NewTags
+func (f *cachingFactory) NewTags(tags []string) *Tags {
+	tagsMap := make(map[uint64]string, len(tags))
+	hash := uint64(0)
+	for _, t := range tags {
+		h := murmur3.StringSum64(t)
+		_, seen := tagsMap[h]
+		if seen {
+			continue
+		}
+		tagsMap[h] = t
+		hash ^= h
+	}
+
+	return f.getCachedTags(byTagsetHashCache, hash, func() *Tags {
+		// write hashes and rewrite tags based on the map
+		hashes := make([]uint64, len(tagsMap))
+		tags = tags[:len(tagsMap)]
+		i := 0
+		for h, t := range tagsMap {
+			tags[i] = t
+			hashes[i] = h
+			i++
+		}
+
+		return &Tags{tags, hashes, hash}
+	})
+}
+
+// NewUniqueTags implements Factory.NewUniqueTags
+func (f *cachingFactory) NewUniqueTags(tags ...string) *Tags {
+	hashes, hash := calcHashes(tags)
+	return f.getCachedTags(byTagsetHashCache, hash, func() *Tags {
+		return &Tags{tags, hashes, hash}
+	})
+}
+
+// NewTagsFromMap implements Factory.NewTagsFromMap
+func (f *cachingFactory) NewTagsFromMap(src map[string]struct{}) *Tags {
+	tags := make([]string, 0, len(src))
+	for tag := range src {
+		tags = append(tags, tag)
+	}
+	hashes, hash := calcHashes(tags)
+	return f.getCachedTags(byTagsetHashCache, hash, func() *Tags {
+		return &Tags{tags, hashes, hash}
+	})
+}
+
+// NewTag implements Factory.NewTag
+func (f *cachingFactory) NewTag(tag string) *Tags {
+	hash := murmur3.StringSum64(tag)
+	return f.getCachedTags(byTagsetHashCache, hash, func() *Tags {
+		return &Tags{[]string{tag}, []uint64{hash}, hash}
+	})
+}
+
+// NewBuilder implements Factory.NewBuilder
+func (f *cachingFactory) NewBuilder(capacity int) *Builder {
+	return f.baseFactory.newBuilder(f, capacity)
+}
+
+// NewSliceBuilder implements Factory.NewSliceBuilder
+func (f *cachingFactory) NewSliceBuilder(levels, capacity int) *SliceBuilder {
+	return f.baseFactory.newSliceBuilder(f, levels, capacity)
+}
+
+// ParseDSD implements Factory.ParseDSD
+func (f *cachingFactory) ParseDSD(data []byte) (*Tags, error) {
+	return f.getCachedTags(byDSDHashCache, murmur3.Sum64(data), func() *Tags {
+		tags := strings.Split(string(data), ",")
+		return f.NewTags(tags)
+	}), nil
+}
+
+// Union implements Factory.Union
+func (f *cachingFactory) Union(a, b *Tags) *Tags {
+	key := unionCacheKey(a.Hash(), b.Hash())
+	return f.getCachedTags(byUnionHashCache, key, func() *Tags {
+		tags := make(map[string]struct{}, len(a.tags)+len(b.tags))
+		for _, t := range a.tags {
+			tags[t] = struct{}{}
+		}
+		for _, t := range b.tags {
+			tags[t] = struct{}{}
+		}
+		return f.NewTagsFromMap(tags)
+	})
+}
+
+// DisjointUnion implements Factory.DisjointUnion
```

Comment:
OK I've just seen in the `Factory` comments that you mention this.

I'm wondering if there is any use-case where the caller will know for sure that there is no duplicates.

This method is slightly dangerous, we should maybe emphase (in the factory comment?) that it is the caller responsibility. I mean, you already did it, but maybe emphase it more, because it could lead to undefined behavior and if the mistake is made by someone, it will very hard to troubleshoot since there is no error message nor warning nor anything (obv. for perf reasons).

Reply:
As noted in another comment, I renamed this to UnsafeDisjointUnion.  The "Unsafe" will serve as a marker that use of this method deserves careful consideration and review.